sdk/appcenter/src/main/java/com/microsoft/appcenter/persistence/DatabasePersistence.java

```diff
@@ -126,18 +153,56 @@ private static ContentValues getContentValues(@Nullable String group, @Nullable
     }
 
     @Override
-    public void putLog(@NonNull String group, @NonNull Log log) throws PersistenceException {
+    public long putLog(@NonNull String group, @NonNull Log log) throws PersistenceException {
 
         /* Convert log to JSON string and put in the database. */
         try {
             AppCenterLog.debug(LOG_TAG, "Storing a log to the Persistence database for log type " + log.getType() + " with sid=" + log.getSid());
-            long databaseId = mDatabaseStorage.put(getContentValues(group, getLogSerializer().serializeLog(log)));
+            String payload = getLogSerializer().serializeLog(log);
+            ContentValues contentValues;
+            boolean isLargePayload = payload.getBytes("UTF-8").length >= PAYLOAD_MAX_SIZE;
+            if (isLargePayload) {
+                contentValues = getContentValues(group, null);
+            } else {
+                contentValues = getContentValues(group, payload);
+            }
+            long databaseId = mDatabaseStorage.put(contentValues);
             AppCenterLog.debug(LOG_TAG, "Stored a log to the Persistence database for log type " + log.getType() + " with databaseId=" + databaseId);
+            if (isLargePayload) {
+                AppCenterLog.debug(LOG_TAG, "Payload is larger than what SQLite supports, storing payload in a separate file.");
+                File directory = getLargePayloadGroupDirectory(group);
+                //noinspection ResultOfMethodCallIgnored we'll get an error anyway at write time.
+                directory.mkdir();
+                File payloadFile = getLargePayloadFile(directory, databaseId);
+                StorageHelper.InternalStorage.write(payloadFile, payload);
+                AppCenterLog.debug(LOG_TAG, "Payload written to " + payloadFile);
+            }
+            return databaseId;
         } catch (JSONException e) {
             throw new PersistenceException("Cannot convert to JSON string", e);
+        } catch (IOException e) {
+            throw new PersistenceException("Cannot save large payload in a file", e);
```

Comment:
Shouldn't we delete the database entry if it fails?

Reply:
It will be deleted at read record time, simpler that way.

Reply:
I agree, it is much simpler. But if the maximum number of logs are already stored, we will lose valid logs because of this.